File: packages/aura_sdk_v1_ts/src/stormEncryptionBindingV1.ts
import { createHash } from "node:crypto";

import type { StormClaim521V1, StormPublicInputs521V1 } from "./stormClaimV1.ts";
import {
  bytesToHexLowerV1,
  concatBytesV1,
  decodeCanonicalFixedHexBytesV1,
  FIELD_ELEMENT_521_BYTE_LEN_V1,
  validateFieldElement521BytesV1,
} from "./stormHash521V1.ts";

export const HASH_LEN_V1 = 32;
export const STORM_ENCRYPTION_BINDING_V1_LEN = HASH_LEN_V1 * 6 + FIELD_ELEMENT_521_BYTE_LEN_V1 * 2;
export const AURA_STORM_ENCRYPTION_BINDING_V1_DOMAIN_SEPARATOR = new TextEncoder().encode(
  "AURA_STORM_ENCRYPTION_BINDING_V1",
);

export type StormEncryptionBindingV1 = {
  stormClaimDigest: Uint8Array;
  traceRoot: Uint8Array;
  finalStateX: Uint8Array;
  finalStateY: Uint8Array;
  contextHash: Uint8Array;
  senderId: Uint8Array;
  receiverId: Uint8Array;
  sessionKeyId: Uint8Array;
};

export function buildStormEncryptionBindingV1(
  lowerLayerClaim: StormClaim521V1,
  lowerLayerPublicInputs: StormPublicInputs521V1,
  stormClaimDigest: Uint8Array,
  senderId: Uint8Array,
  receiverId: Uint8Array,
  sessionKeyId: Uint8Array,
): StormEncryptionBindingV1 {
  return {
    stormClaimDigest: cloneBytes32V1(stormClaimDigest, "stormClaimDigest"),
    traceRoot: decodeCanonicalFixedHexBytesV1(lowerLayerClaim.traceRootHex, HASH_LEN_V1, "traceRootHex"),
    finalStateX: decodeFieldElementBytesV1(lowerLayerClaim.finalState.xHex66Be, "finalState.xHex66Be"),
    finalStateY: decodeFieldElementBytesV1(lowerLayerClaim.finalState.yHex66Be, "finalState.yHex66Be"),
    contextHash: decodeCanonicalFixedHexBytesV1(
      lowerLayerPublicInputs.contextHashHex,
      HASH_LEN_V1,
      "contextHashHex",
    ),
    senderId: cloneBytes32V1(senderId, "senderId"),
    receiverId: cloneBytes32V1(receiverId, "receiverId"),
    sessionKeyId: cloneBytes32V1(sessionKeyId, "sessionKeyId"),
  };
}

export function validateStormEncryptionBindingV1(
  binding: StormEncryptionBindingV1,
): StormEncryptionBindingV1 {
  return {
    stormClaimDigest: cloneBytes32V1(binding.stormClaimDigest, "stormClaimDigest"),
    traceRoot: cloneBytes32V1(binding.traceRoot, "traceRoot"),
    finalStateX: validateFieldElement521BytesV1(binding.finalStateX, "finalStateX"),
    finalStateY: validateFieldElement521BytesV1(binding.finalStateY, "finalStateY"),
    contextHash: cloneBytes32V1(binding.contextHash, "contextHash"),
    senderId: cloneBytes32V1(binding.senderId, "senderId"),
    receiverId: cloneBytes32V1(binding.receiverId, "receiverId"),
    sessionKeyId: cloneBytes32V1(binding.sessionKeyId, "sessionKeyId"),
  };
}

export function encodeStormEncryptionBindingV1(
  binding: StormEncryptionBindingV1,
): Uint8Array {
  const canonical = validateStormEncryptionBindingV1(binding);
  return concatBytesV1(
    canonical.stormClaimDigest,
    canonical.traceRoot,
    canonical.finalStateX,
    canonical.finalStateY,
    canonical.contextHash,
    canonical.senderId,
    canonical.receiverId,
    canonical.sessionKeyId,
  );
}

export function deriveStormEncryptionBindingHashV1(
  binding: StormEncryptionBindingV1,
): Uint8Array {
  return new Uint8Array(
    createHash("sha256")
      .update(AURA_STORM_ENCRYPTION_BINDING_V1_DOMAIN_SEPARATOR)
      .update(encodeStormEncryptionBindingV1(binding))
      .digest(),
  );
}

function cloneBytes32V1(bytes: Uint8Array, fieldName: string): Uint8Array {
  if (!(bytes instanceof Uint8Array) || bytes.length !== HASH_LEN_V1) {
    throw new TypeError(`${fieldName} must be ${HASH_LEN_V1} bytes`);
  }
  return new Uint8Array(bytes);
}

function decodeFieldElementBytesV1(value: string, fieldName: string): Uint8Array {
  return validateFieldElement521BytesV1(
    decodeCanonicalFixedHexBytesV1(value, FIELD_ELEMENT_521_BYTE_LEN_V1, fieldName),
    fieldName,
  );
}

export function stormEncryptionBindingToDebugHexV1(
  binding: StormEncryptionBindingV1,
): Record<string, string> {
  const canonical = validateStormEncryptionBindingV1(binding);
  return {
    stormClaimDigestHex: bytesToHexLowerV1(canonical.stormClaimDigest),
    traceRootHex: bytesToHexLowerV1(canonical.traceRoot),
    finalStateXHex: bytesToHexLowerV1(canonical.finalStateX),
    finalStateYHex: bytesToHexLowerV1(canonical.finalStateY),
    contextHashHex: bytesToHexLowerV1(canonical.contextHash),
    senderIdHex: bytesToHexLowerV1(canonical.senderId),
    receiverIdHex: bytesToHexLowerV1(canonical.receiverId),
    sessionKeyIdHex: bytesToHexLowerV1(canonical.sessionKeyId),
  };
}

File: crates/aura_intent_lineage_v1/src/storm_encryption_binding_v1.rs
//! Canonical storm-native binding material for Aura Session Encryption V1.

use core::fmt;

use crate::{
    FieldElement521V1, StormClaim521V1, StormPublicInputs521V1, HASH_LEN_V1,
    FIELD_ELEMENT_521_BYTE_LEN_V1,
};

pub const STORM_ENCRYPTION_BINDING_V1_LEN: usize =
    HASH_LEN_V1 * 6 + FIELD_ELEMENT_521_BYTE_LEN_V1 * 2;
pub const AURA_STORM_ENCRYPTION_BINDING_V1_DOMAIN_SEPARATOR: &[u8] =
    b"AURA_STORM_ENCRYPTION_BINDING_V1";

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct StormEncryptionBindingV1 {
    pub storm_claim_digest: [u8; HASH_LEN_V1],
    pub trace_root: [u8; HASH_LEN_V1],
    pub final_state_x: [u8; FIELD_ELEMENT_521_BYTE_LEN_V1],
    pub final_state_y: [u8; FIELD_ELEMENT_521_BYTE_LEN_V1],
    pub context_hash: [u8; HASH_LEN_V1],
    pub sender_id: [u8; HASH_LEN_V1],
    pub receiver_id: [u8; HASH_LEN_V1],
    pub session_key_id: [u8; HASH_LEN_V1],
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum StormEncryptionBindingErrorV1 {
    InvalidFinalStateX,
    InvalidFinalStateY,
}

impl fmt::Display for StormEncryptionBindingErrorV1 {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidFinalStateX => {
                write!(f, "invalid storm encryption binding final_state_x")
            }
            Self::InvalidFinalStateY => {
                write!(f, "invalid storm encryption binding final_state_y")
            }
        }
    }
}

impl std::error::Error for StormEncryptionBindingErrorV1 {}

pub fn build_storm_encryption_binding_v1(
    lower_layer_claim: &StormClaim521V1,
    lower_layer_public_inputs: &StormPublicInputs521V1,
    storm_claim_digest: [u8; HASH_LEN_V1],
    sender_id: [u8; HASH_LEN_V1],
    receiver_id: [u8; HASH_LEN_V1],
    session_key_id: [u8; HASH_LEN_V1],
) -> StormEncryptionBindingV1 {
    StormEncryptionBindingV1 {
        storm_claim_digest,
        trace_root: lower_layer_claim.trace_root,
        final_state_x: lower_layer_claim.final_state.x.to_bytes(),
        final_state_y: lower_layer_claim.final_state.y.to_bytes(),
        context_hash: lower_layer_public_inputs.context_hash,
        sender_id,
        receiver_id,
        session_key_id,
    }
}

pub fn encode_storm_encryption_binding_v1(
    binding: &StormEncryptionBindingV1,
) -> Result<[u8; STORM_ENCRYPTION_BINDING_V1_LEN], StormEncryptionBindingErrorV1> {
    validate_storm_encryption_binding_v1(binding)?;

    let mut bytes = [0u8; STORM_ENCRYPTION_BINDING_V1_LEN];
    let mut cursor = 0usize;

    bytes[cursor..cursor + HASH_LEN_V1].copy_from_slice(&binding.storm_claim_digest);
    cursor += HASH_LEN_V1;

    bytes[cursor..cursor + HASH_LEN_V1].copy_from_slice(&binding.trace_root);
    cursor += HASH_LEN_V1;

    bytes[cursor..cursor + FIELD_ELEMENT_521_BYTE_LEN_V1].copy_from_slice(&binding.final_state_x);
    cursor += FIELD_ELEMENT_521_BYTE_LEN_V1;

    bytes[cursor..cursor + FIELD_ELEMENT_521_BYTE_LEN_V1].copy_from_slice(&binding.final_state_y);
    cursor += FIELD_ELEMENT_521_BYTE_LEN_V1;

    bytes[cursor..cursor + HASH_LEN_V1].copy_from_slice(&binding.context_hash);
    cursor += HASH_LEN_V1;

    bytes[cursor..cursor + HASH_LEN_V1].copy_from_slice(&binding.sender_id);
    cursor += HASH_LEN_V1;

    bytes[cursor..cursor + HASH_LEN_V1].copy_from_slice(&binding.receiver_id);
    cursor += HASH_LEN_V1;

    bytes[cursor..cursor + HASH_LEN_V1].copy_from_slice(&binding.session_key_id);

    Ok(bytes)
}

pub fn derive_storm_encryption_binding_hash_v1(
    binding: &StormEncryptionBindingV1,
) -> Result<[u8; HASH_LEN_V1], StormEncryptionBindingErrorV1> {
    let bytes = encode_storm_encryption_binding_v1(binding)?;
    Ok(crate::sha256_domain_separated(
        AURA_STORM_ENCRYPTION_BINDING_V1_DOMAIN_SEPARATOR,
        &bytes,
    ))
}

pub fn validate_storm_encryption_binding_v1(
    binding: &StormEncryptionBindingV1,
) -> Result<(), StormEncryptionBindingErrorV1> {
    FieldElement521V1::from_bytes(binding.final_state_x)
        .map_err(|_| StormEncryptionBindingErrorV1::InvalidFinalStateX)?;
    FieldElement521V1::from_bytes(binding.final_state_y)
        .map_err(|_| StormEncryptionBindingErrorV1::InvalidFinalStateY)?;
    Ok(())
}
